src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs

```diff
@@ -0,0 +1,171 @@
+            if (timeout != Timeout.InfiniteTimeSpan && timeout < TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeout));
+            }
+
+            if (timeProvider is null)
+            {
+                throw new ArgumentNullException(nameof(timeProvider));
+            }
+
+            if (task.IsCompleted)
+            {
+                return task;
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled(cancellationToken);
+            }
+
+            var state = new WaitAsyncState();
+
+            state.Timer = timeProvider.CreateTimer(static s =>
+            {
+                var state = (WaitAsyncState)s!;
+
+                state.TrySetException(new TimeoutException());
+
+                state.Registration.Dispose();
+                state.Timer!.Dispose();
+                state.ContinuationCancellation.Cancel();
+            }, state, Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
+            state.Timer.Change(timeout, Timeout.InfiniteTimeSpan);
+
+            _ = task.ContinueWith(static (t, s) =>
+            {
+                var state = (WaitAsyncState)s!;
+
+                if (t.IsFaulted) state.TrySetException(t.Exception.InnerExceptions);
+                else if (t.IsCanceled) state.TrySetCanceled();
+                else state.TrySetResult(true);
+
+                state.Registration.Dispose();
+                state.Timer?.Dispose();
+            }, state, state.ContinuationCancellation.Token, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
+
+            state.Registration = cancellationToken.Register(static s =>
+            {
+                var state = (WaitAsyncState)s!;
+
+                state.TrySetCanceled();
+
+                state.Timer?.Dispose();
+                state.ContinuationCancellation.Cancel();
+            }, state);
+
+            if (state.Task.IsCompleted)
+            {
+                state.Registration.Dispose();
+            }
+
+            return state.Task;
+        }
+
+        public static Task<TResult> WaitAsync<TResult>(this Task<TResult> task, TimeSpan timeout, TimeProvider timeProvider, CancellationToken cancellationToken = default)
+        {
+            var tcs = new TaskCompletionSource<TResult>();
+
+            WaitAsync((Task)task, timeout, timeProvider, cancellationToken).ConfigureAwait(false).GetAwaiter().OnCompleted(() =>
+            {
+                if (task.IsCompleted)
+                {
+                    tcs.TrySetResult(task.Result);
+                }
+                else if (cancellationToken.IsCancellationRequested)
+                {
+                    tcs.TrySetCanceled(cancellationToken);
+                }
+                else
+                {
+                    tcs.TrySetException(new TimeoutException());
+                }
+            });
+
+            return tcs.Task;
+        }
```

Comment:
```suggestion
        public static async Task<TResult> WaitAsync<TResult>(this Task<TResult> task, TimeSpan timeout, TimeProvider timeProvider, CancellationToken cancellationToken = default)
        {
            await ((Task)task).WaitAsync(timeout, timeProvider, cancellationToken).ConfigureAwait(false);
            return task.Result;
        }
```

Reply:
Depending on how much you care about perf here, you could also share the main implementation.  You'd make it generic on a TResult but still only accept a Task (not a `Task<TResult>`).  The TCS would be generic with `TResult`.  When you go to call TrySetResult, you'd do a type test on the antecedent task and either store `default` if the antecedent task is non-generic or its Result if it is generic.  You should be able to share almost everything and avoid having this extra layer of async method for the generic path.